### DISCUSSION TITLE
How to do cache invalidation on error for AutoDisposeRef and keep alive for data

### DISCUSSION THREAD

**Discussion by GokulDas027** [rrousselGit/riverpod#3172] [Q&A]
Following the docs, i've implemented a timer based cache keep alive for the AutoDisposeRef as an extension, 
but couldn't figure out `how to keep the cache alive only on data and invalidate for error state`, is it possible to access the provider state from inside the extension function on AutoDisposeRef or better alternate ways.

Also, thinking about an optimistic update, like in React query, where invalidated data is shown to user, while fetching the data in the background and replaced once ready. help appreciated.

```
import 'dart:async';

import 'package:hooks_riverpod/hooks_riverpod.dart';

extension RiverpodTimerExtensions on AutoDisposeRef<Object?> {
  /// Keeps the provider alive for [duration] even after all the listeners are removed.
  void keepAliveFor(Duration duration) {
    final link = keepAlive();

    Timer? timer;

    onCancel(() {
      timer = Timer(duration, link.close);
    });

    onResume(() {
      timer?.cancel();
    });

    onDispose(() {
      timer?.cancel();
    });
  }
}
```

**Comment by rrousselGit** [comment]
>  is it possible to access the provider state from inside the extension function on AutoDisposeRef or better alternate ways.

You can read `ref.state`

**Reply by GokulDas027** [reply]
Couldn't access `ref` or `state` directlty, it's reported as undefined. I tried adding to the above extension method.
